content/blog/host-based-volume-encryption-with-hpe-csi-driver-for-kubernetes.md

```diff
@@ -0,0 +1,249 @@
+
+```yaml
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-encrypted-pvc
+  annotations:
+    csi.hpe.com/hostEncryption: "true"
+    csi.hpe.com/hostEncryptionSecretName: my-encryption-key
+    csi.hpe.com/hostEncryptionSecretNamespace: my-namespace
+spec:
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 64Gi
+  storageClassName: hpe-standard-delegated
+```
+
+Let's use the above `StorageClass`, `Secret` and `PersistentVolumeClaim` and attach a workload.
+
+# Provisioning and writing data
+
+In this example I'm just bringing up a sleeping `Pod` and attaching the `PersistentVolumeClaim` named "my-encrypted-pvc" created from the previous section.
+
+```yaml
+---
+kind: Pod
+apiVersion: v1
+metadata:
+  name: my-pod
+spec:
+  containers:
+    - name: my-test-pod
+      image: alpine
+      command: ["/bin/sh"]
+      args: ["-c", "while true; do echo Snoozing...; sleep 10; done"]
+      volumeMounts:
+        - name: my-encrypted-mount
+          mountPath: /data
+  volumes:
+    - name: my-encrypted-mount
+      persistentVolumeClaim:
+        claimName: my-encrypted-pvc
+```
+
+Once the `Pod` has come up, we can inspect the mount point.
+
+```markdown
+$ kubectl exec pod/my-pod -- df -h /data
+Filesystem              Size   Used   Available  Use%  Mounted on
+/dev/mapper/enc-mpatha  64.0G  32.2M  63.9G      0%    /data
+```
+
+The important difference we can observe here versus an unencrypted `PersistentVolumeClaim` is the "enc" keyword in the device name. This means that the kernel module "dm-crypt" has been used to broker access to the actual volume via Linux Unified Key Setup (LUKS). The HPE CSI Driver uses the default cipher with "dm-crypt" which currently is "aes-xts-plain64" with a key size of 256 bits.
+
+# Inspecting the wire
+
+Assume we would write "Hello World" into an unencrypted volume versus an encrypted. In the following experiment I captured `echo "Hello World" > /data/file.txt && sync` with a network packet sniffer and loaded up the result in Wireshark.
+
+![Hello World unencrypted](/img/unecrypted.png "Hello World unencrypted")
+
+Imagine if this would be Personal Identifiable Information (PII) traveling across a compromised network. A security breach could potentially result in an expensive lawsuit if the information end up in the wrong hands. Let's repeat the experiment with an encrypted device.
+
+![Hello World encrypted](/img/encrypted.png "Hello World encrypted")
+
+Inspecting the same packet in the sequence, it's quite obvious the entire payload has been encrypted. This, of course, has impact on the actual storage. Data you know traditionally compress and deduplicate well, all of a sudden present a near 1:1 representation.
+
+I conducted an experiment where I wrote an 8GiB file, filled with zeroes. Under normal circumstances, such a file is nearly invisible to any enterprise array. Not when encrypted.
```

Comment:
Recommend rephrasing the last sentence to: That's not the case when the data is encrypted, though."